src/tsolvers/egraph/Explainer.h

```diff
@@ -0,0 +1,101 @@
+//
+// Created by Martin Blicha on 15.09.20.
+//
+
+#ifndef OPENSMT_EXPLAINER_H
+#define OPENSMT_EXPLAINER_H
+
+#include "EnodeStore.h"
+#include "UFInterpolator.h"
+#include <memory>
+
+class Explainer {
+protected:
+    //
+    // Fast duplicates checking. Cannot be nested !
+    //
+    struct DupChecker;
+    class DuplicateCheckerData {
+        Map<PTRef,int,PTRefHash>    duplicates;                       // Fast duplicate checking
+        int                         dup_count = 0;                    // Current dup token
+        bool                        free = true;
```

Comment:
I am not sure if I understand why is this flag necessary, and why there needs to be 2 layers: `DupChecker` and `DuplicateCheckerData`

Reply:
The conclusion is that DupChecker is a wrapper to ensure automatic proper initialization and release (in the spirit of RAII) and to prevent recursive calls.